File: problems/CSES/Sorting_and_Searching/room_allocation/room_allocation.py
import sys
from heapq import heappush, heappop
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

# Russian:
"""
Этот код полностью логически правильный, но изза особености Python (в особенности его скорости)
это решение не пройдет, и как мне кажеться, найти решение на Python которое пройдет тесты будет
весьма проблематично.
"""

# English:
"""
This code is entirely logically correct, but due to a peculiarity of Python (particularly its speed),
this solution will not pass the tests, and I think it will be quite difficult to find a Python solution 
that will pass the tests.
"""

def main():
    n = int(input())

    customers = []
    for i in range(n):
        a, b = map(int, input().split())
        customers.append((a, b, i))

    customers.sort()

    rooms = []

    allocation_example = [0] * n
    room_count = 0

    for arrival, departure, idx in customers:
        if rooms and rooms[0][0] < arrival:
            last_departure, room_id = heappop(rooms)
        else:
            room_count += 1
            room_id = room_count

        allocation_example[idx] = room_id
        heappush(rooms, (departure, room_id))

    print(room_count)
    print(*allocation_example)

main()

File: problems/CSES/Sorting_and_Searching/room_allocation/room_allocation.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

struct P {
    ll a;
    ll b;
    int i;

    bool operator<(const P& other) const {
        if (a != other.a)
            return a < other.a;
        if (b != other.b)
            return b < other.b;
        return i < other.i;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<P> customers;
    for (int i = 0; i < n; i++) {
        ll a, b;
        cin >> a >> b;
        customers.push_back({a, b, i});
    }

    sort(customers.begin(), customers.end());
    
    priority_queue<pair<ll,ll>, vector<pair<ll,ll>>, greater<pair<ll,ll>>> rooms;

    vector<ll> allocation_example(n, 0);
    ll room_count = 0;

    for (P customer : customers) {
        ll arrival = customer.a;
        ll departure = customer.b;
        ll idx = customer.i;

        ll room_id;

        if (!rooms.empty() && rooms.top().first < arrival) {
            auto [last_departure, free_room] = rooms.top();
            rooms.pop();

            room_id = free_room;
        } else {
            room_count++;
            room_id = room_count;
        }

        allocation_example[idx] = room_id;
        rooms.push({departure, room_id});
    }
    
    cout << room_count << "\n";

    for (ll pos : allocation_example) 
        cout << pos << " ";
    cout << "\n";

    return 0;
}
